### ISSUE TITLE
FR: Enable docker secrets by allowing `*_FILE` environment variables

### ISSUE THREAD

**Issue by whiskeyriver** [tailscale/tailscale#14210]
### What are you trying to do?

I would like to use docker secrets to store my authkey encrypted. The secret is made available to the container as a file, so I would need some way to get the contents of that file to the TS_AUTHKEY variable.

It's common to support `*_FILE` forms of sensitive variables, signaling to the configurator that it needs to read the value from the file.

E.g., the postgres docker image:
```
$ docker run --name some-postgres -e POSTGRES_PASSWORD_FILE=/run/secrets/postgres-passwd -d postgres
```

Alternatively, supporting a credentials file would solve the issue as I could just point to the mounted secret file: `--creds-file /run/secrets/ts_authkey`

### How should we solve this?

`defaultEnv` and `defaultEnvs` could dispatch to a function, `defaultEnvFromFile` when the `name` endswith `_FILE`. Or something.

Like:
```golang
AuthKey: defaultEnvs([]string{"TS_AUTHKEY", "TS_AUTH_KEY", "TS_AUTHKEY_FILE", "TS_AUTH_KEY_FILE"}, ""),
```

and

```golang
// Add this func to handle _FILE vars
func defaultEnvFromFile(name string) (string, bool) {
	if filePath, ok := os.LookupEnv(name); ok {
		data, err := ioutil.ReadFile(filePath)
		if err != nil {
			return "", false
		}
		return string(data), true
	}
	return "", false
}

// Update this func, and maybe others, to dispatch to defaultEnvFromFile
// when the name ends in `_FILE`.
func defaultEnvs(names []string, defVal string) string {
	for _, name := range names {
		if strings.HasSuffix(name, "_FILE") {
			if v, ok := defaultEnvFromFile(name); ok {
				return v
			}
		} else if v, ok := os.LookupEnv(name); ok {
			return v
		}
	}
	return defVal
}
```

### What is the impact of not solving this?

Without using an encrypted secret, the authkey is in plaintext in the process's environment (e.g.: `cat /proc/1398028/environ`).

**Comment by chrishas35** [comment]
Absent the environment variable, I am passing `--auth-key file:/run/secrets/ts_auth_key` in `TS_EXTRA_ARGS`

**Comment by joshka** [comment]
I've implemented this in https://github.com/tailscale/tailscale/pull/16091. Waiting for a review.

**Comment by therealvio** [comment]
Reviewing @joshka's PR would be really sweet - could that please get looked at? I get that there's already argument support using `TS_EXTRA_ARGS` though a dedicated `TS_AUTHKEY_FILE` environment variable does make it easier to do the right thing in terms of safeguarding secrets.

Thanks gang!

**Comment by joshka** [comment]
ping

**Comment by byroncollins** [comment]
I agree with @therealvio . Adding _FILE to variable names is a recognized and secure pattern in Docker for handling secrets. It’s not enforced by Docker, but widely adopted in official images and best practices for secret management

**Comment by joshka** [comment]
I noticed that `containerboot` already supports file-backed secrets by allowing `TS_AUTHKEY` to point at a `file:` path, for example:

```yaml
secrets:
  - ts_authkey
environment:
  TS_AUTHKEY: file:/run/secrets/ts_authkey
```

That does solve the core problem of loading the auth key from a mounted secret.

I still think supporting `TS_AUTHKEY_FILE` is the better interface for this use case, and I’ve implemented it in #16091.

Why I think `_FILE` is still better than `file:` in `TS_AUTHKEY`:

- It matches established container-secret conventions used by many images and deployment tools.
- It keeps the secret value and the indirection mechanism separate instead of overloading `TS_AUTHKEY`.
- It is easier to discover and easier to wire through Compose, Helm, Kubernetes, and secret-injection tooling.
- It avoids relying on an app-specific `file:` prefix embedded in a value field that normally contains the credential itself.

So this is less about enabling file-backed secrets for the first time, and more about exposing that capability through a more standard and operationally clearer interface.

**Comment by joshka** [comment]
Some discussion of this on Reddit at https://www.reddit.com/r/Tailscale/comments/1tlk328/comment/ongfr01/
